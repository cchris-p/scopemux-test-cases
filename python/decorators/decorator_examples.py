#!/usr/bin/env python3
"""
Demonstration of Python decorators

This example shows:
- Basic function decorators
- Decorators with arguments
- Class decorators
- Method decorators
- Decorator stacking
- Functools wraps to preserve metadata
- Creating decorators with classes
"""

import functools
import time
from typing import Callable, Any, TypeVar, cast


# Basic function decorator
def simple_decorator(func: Callable) -> Callable:
    """A simple decorator that prints before and after function execution"""

    def wrapper(*args, **kwargs):
        print(f"Before calling {func.__name__}")
        result = func(*args, **kwargs)
        print(f"After calling {func.__name__}")
        return result

    return wrapper


# Using the simple decorator
@simple_decorator
def greet(name: str) -> str:
    """Return a greeting message"""
    return f"Hello, {name}!"


# Decorator with arguments
def repeat(times: int) -> Callable:
    """A decorator that repeats the function call a specified number of times"""

    def decorator(func: Callable) -> Callable:
        def wrapper(*args, **kwargs):
            results = []
            for _ in range(times):
                results.append(func(*args, **kwargs))
            return results

        return wrapper

    return decorator


# Using the decorator with arguments
@repeat(3)
def say_hi(name: str) -> str:
    """Say hi to someone"""
    return f"Hi, {name}!"


# Using functools.wraps to preserve metadata
def timing_decorator(func: Callable) -> Callable:
    """A decorator that times function execution"""

    @functools.wraps(func)  # Preserves the original function's metadata
    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        print(f"{func.__name__} took {end_time - start_time:.4f} seconds to run")
        return result

    return wrapper


# Using the timing decorator
@timing_decorator
def slow_function() -> None:
    """A function that takes some time to execute"""
    time.sleep(0.5)
    print("Slow function executed")


# Stacking decorators
@simple_decorator
@timing_decorator
def combined_decorators() -> None:
    """Function with multiple decorators"""
    print("Function with stacked decorators")
    time.sleep(0.1)


# Class decorator
def add_greeting(cls):
    """A class decorator that adds a greeting method"""

    def say_hello(self):
        return f"Hello from {self.__class__.__name__}!"

    cls.say_hello = say_hello
    return cls


# Using the class decorator
@add_greeting
class Person:
    """A simple person class"""

    def __init__(self, name: str, age: int):
        self.name = name
        self.age = age

    def __str__(self) -> str:
        return f"{self.name}, {self.age} years old"


# Method decorator
def validate_age(method: Callable) -> Callable:
    """A decorator to validate age parameter"""

    @functools.wraps(method)
    def wrapper(self, age, *args, **kwargs):
        if age < 0 or age > 120:
            raise ValueError("Age must be between 0 and 120")
        return method(self, age, *args, **kwargs)

    return wrapper


# Class with a decorated method
class Student:
    """A student class with a decorated method"""

    def __init__(self, name: str, age: int):
        self.name = name
        self.set_age(age)

    @validate_age
    def set_age(self, age: int) -> None:
        """Set student age with validation"""
        self.age = age

    def __str__(self) -> str:
        return f"Student: {self.name}, {self.age} years old"


# Creating a decorator using a class
class CountCalls:
    """A decorator implemented as a class that counts function calls"""

    def __init__(self, func: Callable):
        functools.update_wrapper(self, func)
        self.func = func
        self.count = 0

    def __call__(self, *args, **kwargs):
        self.count += 1
        print(f"{self.func.__name__} has been called {self.count} times")
        return self.func(*args, **kwargs)


# Using the class-based decorator
@CountCalls
def increment(x: int) -> int:
    """Increment a number"""
    return x + 1


# Generic decorator with TypeVar
T = TypeVar("T")


def debug(func: Callable[..., T]) -> Callable[..., T]:
    """Print function arguments and return value"""

    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        args_repr = [repr(a) for a in args]
        kwargs_repr = [f"{k}={v!r}" for k, v in kwargs.items()]
        signature = ", ".join(args_repr + kwargs_repr)
        print(f"Calling {func.__name__}({signature})")
        result = func(*args, **kwargs)
        print(f"{func.__name__} returned {result!r}")
        return result

    return cast(Callable[..., T], wrapper)


# Using the debug decorator
@debug
def divide(a: float, b: float) -> float:
    """Divide two numbers"""
    return a / b


def main():
    """Main function to demonstrate decorators"""
    print("=== Basic Decorator ===")
    message = greet("Alice")
    print(f"Result: {message}")

    print("\n=== Decorator with Arguments ===")
    results = say_hi("Bob")
    print(f"Results: {results}")

    print("\n=== Timing Decorator ===")
    slow_function()

    print("\n=== Stacked Decorators ===")
    combined_decorators()

    print("\n=== Class Decorator ===")
    person = Person("Charlie", 30)
    print(person)
    print(person.say_hello())

    print("\n=== Method Decorator ===")
    student = Student("David", 20)
    print(student)
    try:
        invalid_student = Student("Eve", 150)  # Should raise an error
    except ValueError as e:
        print(f"Validation error: {e}")

    print("\n=== Class-Based Decorator ===")
    print(increment(5))
    print(increment(10))
    print(increment(15))

    print("\n=== Debug Decorator ===")
    result = divide(10, 2)
    print(f"Final result: {result}")


if __name__ == "__main__":
    main()
