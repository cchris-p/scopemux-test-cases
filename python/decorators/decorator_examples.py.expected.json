{
  "language": "Python",
  "ast": {
    "type": "module",
    "functions": [
      {
        "name": "simple_decorator",
        "type": "function_definition",
        "parameters": [
          {"name": "func", "type": "Callable"}
        ],
        "return_type": "Callable",
        "body_range": {
          "start_line": 16,
          "start_column": 0,
          "end_line": 25,
          "end_column": 4
        },
        "docstring": "A simple decorator that prints before and after function execution"
      },
      {
        "name": "greet",
        "type": "function_definition",
        "parameters": [
          {"name": "name", "type": "str"}
        ],
        "return_type": "str",
        "body_range": {
          "start_line": 29,
          "start_column": 0,
          "end_line": 31,
          "end_column": 4
        },
        "docstring": "Return a greeting message",
        "decorators": ["simple_decorator"]
      },
      {
        "name": "repeat",
        "type": "function_definition",
        "parameters": [
          {"name": "times", "type": "int"}
        ],
        "return_type": "Callable",
        "body_range": {
          "start_line": 34,
          "start_column": 0,
          "end_line": 46,
          "end_column": 4
        },
        "docstring": "A decorator that repeats the function call a specified number of times"
      },
      {
        "name": "say_hi",
        "type": "function_definition",
        "parameters": [
          {"name": "name", "type": "str"}
        ],
        "return_type": "str",
        "body_range": {
          "start_line": 50,
          "start_column": 0,
          "end_line": 52,
          "end_column": 4
        },
        "docstring": "Say hi to someone",
        "decorators": ["repeat(3)"]
      },
      {
        "name": "timing_decorator",
        "type": "function_definition",
        "parameters": [
          {"name": "func", "type": "Callable"}
        ],
        "return_type": "Callable",
        "body_range": {
          "start_line": 55,
          "start_column": 0,
          "end_line": 65,
          "end_column": 4
        },
        "docstring": "A decorator that times function execution"
      },
      {
        "name": "slow_function",
        "type": "function_definition",
        "parameters": [],
        "return_type": "None",
        "body_range": {
          "start_line": 69,
          "start_column": 0,
          "end_line": 72,
          "end_column": 4
        },
        "docstring": "A function that takes some time to execute",
        "decorators": ["timing_decorator"]
      },
      {
        "name": "combined_decorators",
        "type": "function_definition",
        "parameters": [],
        "return_type": "None",
        "body_range": {
          "start_line": 77,
          "start_column": 0,
          "end_line": 79,
          "end_column": 4
        },
        "docstring": "Function with multiple decorators",
        "decorators": ["simple_decorator", "timing_decorator"]
      },
      {
        "name": "add_greeting",
        "type": "function_definition",
        "parameters": [
          {"name": "cls", "type": null}
        ],
        "body_range": {
          "start_line": 82,
          "start_column": 0,
          "end_line": 89,
          "end_column": 4
        },
        "docstring": "A class decorator that adds a greeting method"
      },
      {
        "name": "validate_age",
        "type": "function_definition",
        "parameters": [
          {"name": "method", "type": "Callable"}
        ],
        "return_type": "Callable",
        "body_range": {
          "start_line": 102,
          "start_column": 0,
          "end_line": 111,
          "end_column": 4
        },
        "docstring": "A decorator to validate age parameter"
      },
      {
        "name": "debug",
        "type": "function_definition",
        "parameters": [
          {"name": "func", "type": "Callable[..., T]"}
        ],
        "return_type": "Callable[..., T]",
        "body_range": {
          "start_line": 150,
          "start_column": 0,
          "end_line": 163,
          "end_column": 4
        },
        "docstring": "Print function arguments and return value"
      },
      {
        "name": "divide",
        "type": "function_definition",
        "parameters": [
          {"name": "a", "type": "float"},
          {"name": "b", "type": "float"}
        ],
        "return_type": "float",
        "body_range": {
          "start_line": 167,
          "start_column": 0,
          "end_line": 169,
          "end_column": 4
        },
        "docstring": "Divide two numbers",
        "decorators": ["debug"]
      },
      {
        "name": "main",
        "type": "function_definition",
        "parameters": [],
        "body_range": {
          "start_line": 172,
          "start_column": 0,
          "end_line": 204,
          "end_column": 4
        },
        "docstring": "Main function to demonstrate decorators"
      }
    ],
    "classes": [
      {
        "name": "Person",
        "type": "class_definition",
        "methods": [
          {
            "name": "__init__",
            "type": "method",
            "parameters": ["self", "name: str", "age: int"],
            "body_range": {
              "start_line": 95,
              "start_column": 4,
              "end_line": 97,
              "end_column": 8
            }
          },
          {
            "name": "__str__",
            "type": "method",
            "parameters": ["self"],
            "return_type": "str",
            "body_range": {
              "start_line": 99,
              "start_column": 4,
              "end_line": 100,
              "end_column": 8
            }
          },
          {
            "name": "say_hello",
            "type": "method",
            "parameters": ["self"],
            "added_by_decorator": true
          }
        ],
        "decorators": ["add_greeting"],
        "range": {
          "start_line": 93,
          "start_column": 0,
          "end_line": 100,
          "end_column": 51
        },
        "docstring": "A simple person class"
      },
      {
        "name": "Student",
        "type": "class_definition",
        "methods": [
          {
            "name": "__init__",
            "type": "method",
            "parameters": ["self", "name: str", "age: int"],
            "body_range": {
              "start_line": 115,
              "start_column": 4,
              "end_line": 117,
              "end_column": 8
            }
          },
          {
            "name": "set_age",
            "type": "method",
            "parameters": ["self", "age: int"],
            "return_type": "None",
            "body_range": {
              "start_line": 120,
              "start_column": 4,
              "end_line": 122,
              "end_column": 8
            },
            "docstring": "Set student age with validation",
            "decorators": ["validate_age"]
          },
          {
            "name": "__str__",
            "type": "method",
            "parameters": ["self"],
            "return_type": "str",
            "body_range": {
              "start_line": 124,
              "start_column": 4,
              "end_line": 125,
              "end_column": 8
            }
          }
        ],
        "range": {
          "start_line": 113,
          "start_column": 0,
          "end_line": 125,
          "end_column": 59
        },
        "docstring": "A student class with a decorated method"
      },
      {
        "name": "CountCalls",
        "type": "class_definition",
        "methods": [
          {
            "name": "__init__",
            "type": "method",
            "parameters": ["self", "func: Callable"],
            "body_range": {
              "start_line": 131,
              "start_column": 4,
              "end_line": 134,
              "end_column": 8
            }
          },
          {
            "name": "__call__",
            "type": "method",
            "parameters": ["self", "*args", "**kwargs"],
            "body_range": {
              "start_line": 136,
              "start_column": 4,
              "end_line": 139,
              "end_column": 8
            }
          }
        ],
        "range": {
          "start_line": 128,
          "start_column": 0,
          "end_line": 139,
          "end_column": 8
        },
        "docstring": "A decorator implemented as a class that counts function calls"
      }
    ],
    "imports": [
      {
        "name": "functools",
        "line": 11
      },
      {
        "name": "time",
        "line": 12
      },
      {
        "name": "typing",
        "symbols": ["Callable", "Any", "TypeVar", "cast"],
        "line": 13
      }
    ],
    "docstrings": [
      {
        "type": "module_docstring",
        "text": "Demonstration of Python decorators\n\nThis example shows:\n- Basic function decorators\n- Decorators with arguments\n- Class decorators\n- Method decorators\n- Decorator stacking\n- Functools wraps to preserve metadata\n- Creating decorators with classes",
        "range": {
          "start_line": 1,
          "start_column": 0,
          "end_line": 10,
          "end_column": 3
        }
      }
    ],
    "decorators": [
      {
        "name": "simple_decorator",
        "applied_to": "greet",
        "line": 28
      },
      {
        "name": "repeat(3)",
        "applied_to": "say_hi",
        "line": 49
      },
      {
        "name": "functools.wraps",
        "applied_to": "wrapper",
        "line": 58
      },
      {
        "name": "timing_decorator",
        "applied_to": "slow_function",
        "line": 68
      },
      {
        "name": "simple_decorator",
        "applied_to": "combined_decorators",
        "line": 75
      },
      {
        "name": "timing_decorator",
        "applied_to": "combined_decorators",
        "line": 76
      },
      {
        "name": "add_greeting",
        "applied_to": "Person",
        "line": 92
      },
      {
        "name": "functools.wraps",
        "applied_to": "wrapper",
        "line": 105
      },
      {
        "name": "validate_age",
        "applied_to": "set_age",
        "line": 119
      },
      {
        "name": "CountCalls",
        "applied_to": "increment",
        "line": 142
      },
      {
        "name": "functools.wraps",
        "applied_to": "wrapper",
        "line": 153
      },
      {
        "name": "debug",
        "applied_to": "divide",
        "line": 166
      }
    ],
    "variable_declarations": [
      {
        "name": "T",
        "type": "TypeVar",
        "line": 148,
        "initialized": true
      }
    ],
    "function_calls": [
      {
        "name": "main",
        "arguments": [],
        "line": 207
      }
    ],
    "type_annotations": [
      {
        "name": "func",
        "type": "Callable",
        "line": 16
      },
      {
        "name": "simple_decorator",
        "return_type": "Callable",
        "line": 16
      },
      {
        "name": "name",
        "type": "str",
        "line": 29
      },
      {
        "name": "greet",
        "return_type": "str",
        "line": 29
      },
      {
        "name": "times",
        "type": "int",
        "line": 34
      },
      {
        "name": "repeat",
        "return_type": "Callable",
        "line": 34
      },
      {
        "name": "name",
        "type": "str",
        "line": 50
      },
      {
        "name": "say_hi",
        "return_type": "str",
        "line": 50
      },
      {
        "name": "func",
        "type": "Callable",
        "line": 55
      },
      {
        "name": "timing_decorator",
        "return_type": "Callable",
        "line": 55
      },
      {
        "name": "slow_function",
        "return_type": "None",
        "line": 69
      },
      {
        "name": "combined_decorators",
        "return_type": "None",
        "line": 77
      },
      {
        "name": "method",
        "type": "Callable",
        "line": 102
      },
      {
        "name": "validate_age",
        "return_type": "Callable",
        "line": 102
      },
      {
        "name": "name",
        "type": "str",
        "line": 95
      },
      {
        "name": "age",
        "type": "int",
        "line": 95
      },
      {
        "name": "__str__",
        "return_type": "str",
        "line": 99
      },
      {
        "name": "name",
        "type": "str",
        "line": 115
      },
      {
        "name": "age",
        "type": "int",
        "line": 115
      },
      {
        "name": "age",
        "type": "int",
        "line": 120
      },
      {
        "name": "set_age",
        "return_type": "None",
        "line": 120
      },
      {
        "name": "__str__",
        "return_type": "str",
        "line": 124
      },
      {
        "name": "func",
        "type": "Callable",
        "line": 131
      },
      {
        "name": "x",
        "type": "int",
        "line": 143
      },
      {
        "name": "increment",
        "return_type": "int",
        "line": 143
      },
      {
        "name": "func",
        "type": "Callable[..., T]",
        "line": 150
      },
      {
        "name": "debug",
        "return_type": "Callable[..., T]",
        "line": 150
      },
      {
        "name": "a",
        "type": "float",
        "line": 167
      },
      {
        "name": "b",
        "type": "float",
        "line": 167
      },
      {
        "name": "divide",
        "return_type": "float",
        "line": 167
      }
    ]
  }
}
